core/primitives/src/transaction.rs

```diff
@@ -210,6 +216,90 @@ impl From<DeleteAccountAction> for Action {
     }
 }
 
+#[derive(Serialize, BorshSerialize, Deserialize, PartialEq, Eq, Clone, Debug)]
+pub struct NonDelegateAction(pub Action);
+
+impl From<NonDelegateAction> for Action {
+    fn from(action: NonDelegateAction) -> Self {
+        action.0
+    }
+}
+
+impl borsh::de::BorshDeserialize for NonDelegateAction {
+    fn deserialize(buf: &mut &[u8]) -> ::core::result::Result<Self, borsh::maybestd::io::Error> {
+        if buf.is_empty() {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "Failed to deserialize DelegateAction",
+            ));
+        }
+        match buf[0] {
+            ACTION_DELEGATE_NUMBER => Err(Error::new(
+                ErrorKind::InvalidInput,
+                "DelegateAction mustn't contain a nested one",
+            )),
+            _ => Ok(Self(borsh::BorshDeserialize::deserialize(buf)?)),
+        }
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, PartialEq, Eq, Clone, Debug)]
+pub struct DelegateAction {
+    /// Signer of the delegated actions
+    pub sender_id: AccountId,
+    /// Receiver of the delegated actions.
+    pub receiver_id: AccountId,
+    /// List of actions to be executed.
+    pub actions: Vec<NonDelegateAction>,
+    /// Nonce to ensure that the same delegate action is not sent twice by a relayer and should match for given account's `public_key`.
+    /// After this action is processed it will increment.
+    pub nonce: Nonce,
+    /// The maximal height of the block in the blockchain below which the given DelegateAction is valid.
+    pub max_block_height: BlockHeight,
+    /// Public key that is used to sign this delegated action.
+    pub public_key: PublicKey,
+}
+#[cfg_attr(feature = "protocol_feature_delegate_action", derive(BorshDeserialize))]
+#[derive(BorshSerialize, Serialize, Deserialize, PartialEq, Eq, Clone, Debug)]
+pub struct SignedDelegateAction {
+    pub delegate_action: DelegateAction,
+    pub signature: Signature,
+}
+
+#[cfg(not(feature = "protocol_feature_delegate_action"))]
+impl borsh::de::BorshDeserialize for SignedDelegateAction {
+    fn deserialize(_buf: &mut &[u8]) -> ::core::result::Result<Self, borsh::maybestd::io::Error> {
+        return Err(Error::new(ErrorKind::InvalidInput, "Delegate action isn't supported"));
+    }
+}
+
+impl SignedDelegateAction {
+    pub fn verify(&self) -> bool {
+        let delegate_action = &self.delegate_action;
+        let hash = delegate_action.get_hash();
+        let public_key = &delegate_action.public_key;
+
+        self.signature.verify(hash.as_ref(), public_key)
+    }
+}
+
+impl From<SignedDelegateAction> for Action {
+    fn from(delegate_action: SignedDelegateAction) -> Self {
+        Self::Delegate(delegate_action)
+    }
+}
+
+impl DelegateAction {
+    pub fn get_actions(&self) -> Vec<Action> {
```

Comment:
you should be able to avoid the clone by referencing the inner actions, no?

Reply:
Yes I can return `Vec<&Action>` but there is no sense, I think. There are only three cases when `get_actions` is used:
1. When the returned `Vec<Action>` is passed to another function in config.rs as an `&[Action]` argument. So cloning `&Action` is required in this case.
2. When  a new receipt is created in actions.rs. In this case, we have to copy `Vec<Action>` in any case.
3. In `validate_delegate_action_key`. This is the only case when `Vec<&Action>` would be helpful.